### ISSUE TITLE
Feature: IterableDataset support for SFTTrainer

### ISSUE THREAD

**Issue by yash4242** [huggingface/trl#1695]
Hi,
I've been trying to finetune a language model on a standard dataset, with `streaming=True`, i.e. the dataset thus made is an `IterableDataset`. How do I feed a dataset in such a streaming mode to the `SFTTrainer` (and/or `Trainer.train()`)?

### Versions:
```
Python 3.9.12 (main, Jun  1 2022, 11:38:51) 
>>> transformers.__version__
'4.40.2'
>>> datasets.__version__
'2.19.2'
>>> trl.__version__
'0.8.6'
```
OS: RHEL

### Usecase:
have a dataset in streaming mode, and do supervised finetuning on it in the streaming mode.

### Problems faced / Steps to reproduce:
```py
iterable_dataset = load_dataset("imdb", split="train", streaming=True)
trainer = SFTTrainer(
    model=model, # defined somewhere else, not relevant to this issue
    args=training_args, #  defined somewhere else, not relevant to this issue
    train_dataset=iterable_dataset,
    max_seq_length=512,
    dataset_text_field="text",
    tokenizer=tokenizer, #  defined somewhere else, not relevant to this issue
    )
```
**Error Thrown:**
`TypeError: map() got an unexpected keyword argument 'num_proc'`

**Complete Error Trace:**

```
Traceback (most recent call last):
  File "/u/yash/sft.py", line 75, in <module>
    main(args)
  File "/u/yash/sft.py", line 50, in main
    trainer = SFTTrainer(
  File "/u/yash/.conda/envs/myenv/lib/python3.9/site-packages/trl/trainer/sft_trainer.py", line 283, in __init__
    train_dataset = self._prepare_dataset(
  File "/u/yash/.conda/envs/myenv/lib/python3.9/site-packages/trl/trainer/sft_trainer.py", line 426, in _prepare_dataset
    return self._prepare_non_packed_dataloader(
  File "/u/yash/.conda/envs/myenv/lib/python3.9/site-packages/trl/trainer/sft_trainer.py", line 493, in _prepare_non_packed_dataloader
    tokenized_dataset = dataset.map(
TypeError: map() got an unexpected keyword argument 'num_proc'
```
The problem is that `iterable_dataset` here is of `datasets.iterable_dataset.IterableDataset` class. But in the library code `python3.9/site-packages/datasets/iterable_dataset.py` line 1581:
`def map(` definition does not have any argument `num_proc`.

### Workarounds I tried
1.  Just adding the `num_proc: int = None` argument in the function signature. This also does not help because later an error for an undefined `__len__` comes up. Complete error trace:
```
Traceback (most recent call last):
  File "/u/yash/time-prober/sft.py", line 75, in <module>
    main(args)
  File "/u/yash/time-prober/sft.py", line 50, in main
    trainer = SFTTrainer(
  File "/u/yash/.conda/envs/myenv/lib/python3.9/site-packages/trl/trainer/sft_trainer.py", line 323, in __init__
    super().__init__(
  File "/u/yash/.conda/envs/myenv/lib/python3.9/site-packages/transformers/trainer.py", line 581, in __init__
    raise ValueError(
ValueError: The train_dataset does not implement __len__, max_steps has to be specified. The number of steps needs to be known in advance for the learning rate scheduler
```
Which is to be expected because `IterableDataset` class has no `__len__` defined intentionally `python3.9/site-packages/torch/utils/data/dataset.py` line 187.
2. Inheriting this `IterableDataset` class just to implement a `def __len__` which returns a hardcoded value.
3. Found a few fixes on the forums [here](https://discuss.huggingface.co/t/how-to-use-huggingface-trainer-streaming-datasets-without-wrapping-it-with-torchdatas-iterablewrapper/25230), about using `IterableWrapper`. That might work for `Seq2Seq` trainer, but that doesn't work with SFTTrainer. (I need to focus on SFTtrainer, so didn't test it with Seq2Seq trainer). 
4. Another fix as to use `.with_format("torch")`, which also doesnt work for me: 
```py
dataset = load_dataset("imdb", split="train", streaming=True)
iterable_dataset = dataset.with_format("torch")
trainer = SFTTrainer(
    model=model, # defined somewhere else, not relevant to this issue
    args=training_args, #  defined somewhere else, not relevant to this issue
    train_dataset=iterable_dataset,
    max_seq_length=512,
    dataset_text_field="text",
    tokenizer=tokenizer, #  defined somewhere else, not relevant to this issue
    )
```
Error Trace: (exactly the same as without `.with_format("torch")`)
```Traceback (most recent call last):
  File "/u/yash/sft.py", line 75, in <module>
    main(args)
  File "/u/yash/sft.py", line 50, in main
    trainer = SFTTrainer(
  File "/u/yash/.conda/envs/myenv/lib/python3.9/site-packages/trl/trainer/sft_trainer.py", line 283, in __init__
    train_dataset = self._prepare_dataset(
  File "/u/yash/.conda/envs/myenv/lib/python3.9/site-packages/trl/trainer/sft_trainer.py", line 426, in _prepare_dataset
    return self._prepare_non_packed_dataloader(
  File "/u/yash/.conda/envs/myenv/lib/python3.9/site-packages/trl/trainer/sft_trainer.py", line 493, in _prepare_non_packed_dataloader
    tokenized_dataset = dataset.map(
TypeError: map() got an unexpected keyword argument 'num_proc'
```
5. Using `.take` repeatedly to request a chunk of data, and feeding that to the trainer. I haven't tried this out. Is it possible to switch datasets/parts of a dataset within one epoch with huggingface's trainer?

Apologies if there's some fix already available which I missed. I will be grateful if you could share it. If IterableDataset support is really not there for Trainer, I would be happy to contribute!

Thanks!

**Comment by ironbar** [comment]
Why this issue was closed? I'm on version `0.9.6` and I get the exact same error as reported here when trying to use `IterableDataset`

**Comment by ironbar** [comment]
It seems it has been fixed, but no release has been created since. Thus, it can be solved by installing directly from github:

```
pip install git+https://github.com/huggingface/trl.git
```
